### ISSUE TITLE
Angular 9 - ngOnDestroy called on not OnDestroy instance when in multi provision

### ISSUE THREAD

**Issue by Azbesciak** [angular/angular#37481]
# 🐞 bug report

### Affected Package
probably `core`, 9.1.9

### Is this a regression?
since 8 - yes, since 9.0 - maybe.
Connected issue - https://github.com/angular/angular/issues/35231
As least I just migrated from 8.2.14

### Description
So, as in mentioned ticket 35231 the problem occurs when `ngOnDestroy` is called.
Recently the context was changed, now it seems to be called on every instance in `multi` provision. Even when some of them does not implement `OnDestroy` (at least it looks so, but as you later will see - it does not fit to logic, context is changed again).
Screenshot from chrome debugger when destroy was called (file `core.js`).
![image](https://user-images.githubusercontent.com/17334466/83994077-a859e780-a955-11ea-8396-ba6ee6137626.png)
The problem is - `RouteResultExternalLinksManager` does not implement `OnDestroy`, and because of that exception occurs:
![image](https://user-images.githubusercontent.com/17334466/83994245-13a3b980-a956-11ea-97cb-5fe8ee8f5781.png)
I would even say that `RouteResultExternalLinksManager` is not an instance of `Subscriber` as it looks like in debugger:
![image](https://user-images.githubusercontent.com/17334466/83994288-36ce6900-a956-11ea-8bd7-a192da8548a2.png)
![image](https://user-images.githubusercontent.com/17334466/83994304-3df57700-a956-11ea-9816-acae3bfa1000.png)

## 🔥 Exception or Error
<pre><code>
TypeError: this.clearSubscriptions is not a function
    at RouteResultExternalLinksManager.push.../../libs/imap/shared/flow-utils/subscriber.ts.Subscriber.ngOnDestroy (subscriber.ts:9)
    at executeOnDestroys (core.js:9522)
    at cleanUpView (core.js:9448)
    at destroyViewTree (core.js:9248)
    at destroyLView (core.js:9401)
    at RootViewRef.push.../../node_modules/@angular/core/__ivy_ngcc__/fesm5/core.js.ViewRef.destroy (core.js:10143)
    at ComponentRef.push.../../node_modules/@angular/core/__ivy_ngcc__/fesm5/core.js.ComponentRef.destroy (core.js:23229)
    at RouterOutlet.push.../../node_modules/@angular/router/__ivy_ngcc__/fesm5/router.js.RouterOutlet.deactivate (router.js:5355)
    at ActivateRoutes.push.../../node_modules/@angular/router/__ivy_ngcc__/fesm5/router.js.ActivateRoutes.deactivateRouteAndOutlet (router.js:2328)
    at ActivateRoutes.push.../../node_modules/@angular/router/__ivy_ngcc__/fesm5/router.js.ActivateRoutes.deactivateRouteAndItsChildren (router.js:2308)

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.1.7
Node: 12.13.1
OS: win32 x64

Angular: 9.1.9
... animations, common, compiler, compiler-cli, core, elements
... forms, language-service, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.901.7
@angular-devkit/build-angular      0.901.7
@angular-devkit/build-ng-packagr   0.901.7
@angular-devkit/build-optimizer    0.901.7
@angular-devkit/build-webpack      0.901.7
@angular-devkit/core               8.3.23
@angular-devkit/schematics         8.3.23
@angular/cdk                       9.2.4
@angular/cli                       9.1.7
@angular/material                  9.2.4
@ngtools/webpack                   9.1.7
@schematics/angular                8.3.23
@schematics/update                 0.901.7
ng-packagr                         9.1.5
rxjs                               6.5.5
typescript                         3.8.3
webpack                            4.42.0
</code></pre>

**Anything else relevant?**
I am using chrome 83.0.4103.97 64 bit if it helps

**Comment by crisbeto** [comment]
@Azbesciak can you post what your multi provider looks like?

**Comment by Azbesciak** [comment]
If you are talking about interface - it is via `@Inject` and `InjectionToken`. 
To provide I am using `provideResultProcessor` like below.
All is provided in component scope, which is a base in a route.
```ts

export const RESULT_PROCESSOR_TOKEN = new InjectionToken("result_processor");

export interface RouteResultProcessor {
  readonly priority: number

  onNewResult(result: RouteResultsAccess): RouteResultsAccess
}

export function provideResultProcessor(type) {
  return {
    provide: RESULT_PROCESSOR_TOKEN,
    useClass: type,
    multi: true
  }
}
```

**Comment by crisbeto** [comment]
Can you post an example where the `provideResultProcessor` is used? I'm trying to put this into a unit test, but it's behaving as expected. Here's what I'm doing:

```
it('should fail', () => {
  const destroyCalls: any[] = [];
  const SERVICES = new InjectionToken<any>('SERVICES');

  @Injectable()
  class Service0 {
  }

  @Injectable()
  class Service1 {
    ngOnDestroy() {
      destroyCalls.push(this);
    }
  }

  @Injectable()
  class Service2 {
  }

  @Injectable()
  class Service3 {
    ngOnDestroy() {
      destroyCalls.push(this);
    }
  }

  @Injectable()
  class Service4 {
  }

  @Component({
    template: '<div></div>',
    providers: [
      {provide: SERVICES, useClass: Service0, multi: true},
      {provide: SERVICES, useClass: Service1, multi: true},
      {provide: SERVICES, useClass: Service2, multi: true},
      {provide: SERVICES, useClass: Service3, multi: true},
      {provide: SERVICES, useClass: Service4, multi: true},
    ]
  })
  class App {
    constructor(@Inject(SERVICES) s: any) {}
  }

  TestBed.configureTestingModule({declarations: [App]});
  const fixture = TestBed.createComponent(App);
  fixture.detectChanges();
  fixture.destroy();

  // Logs [ Service1 {}, Service3 {} ] as expected.
  console.log(destroyCalls);
});
```

**Comment by Azbesciak** [comment]
Sorry for the long delay;
first of all - this provision is in lazy module, in component scope
Processors are used in service created at the same level (in component provision).

As I said `slice` on Array before sorting solves the problem, so it lies probably there.

**Comment by JoostK** [comment]
> PS: those are later sorted via priority - and as I see when I make a slice on that array and sort it - it works then without exception.
BUT that this is strange anyway.

AH! That sounds exactly where the problem would originate. Looks like sorting the array of multi-providers ends up modifying the internal DI record of those providers, such that other internal stuff starts to interpret the data incorrectly.

**Comment by crisbeto** [comment]
That totally makes sense @JoostK! We should probably clone the array before we save it in our internal data structures.

**Comment by crisbeto** [comment]
I took a better look and it and it seems like this uncovered an issue with `multi` providers that has been there since ViewEngine. If the array returned from a `multi` provider is mutated, it'll propagate both in our internal data structures, causing issues like `ngOnDestroy` getting called with the wrong context, as well as any other places that inject the same token. You probably didn't see it, because we weren't calling `ngOnDestroy` for `multi` providers. 

Here's a test that shows the issue. Two instances of the same directive request an injection token and get the providers in a different order:

```it('', () => {
  const instances: any[] = [];
  const SERVICES = new InjectionToken<any>('SERVICES');

  @Injectable()
  class Service0 {
  }

  @Injectable()
  class Service1 {
  }

  @Directive({selector: '[dir]'})
  class Dir {
    constructor(@Inject(SERVICES) s: any[]) {
      instances.push(s.slice());
      s.reverse();
    }
  }

  @Component({
    template: '<div dir></div><div dir></div>',
    providers: [
      {provide: SERVICES, useClass: Service0, multi: true},
      {provide: SERVICES, useClass: Service1, multi: true},
    ]
  })
  class App {
  }

  TestBed.configureTestingModule({declarations: [App, Dir]});
  const fixture = TestBed.createComponent(App);
  fixture.detectChanges();

  // Logs [[Service0, Service1], [Service1, Sevice0]]
  console.log(instances);
});
```

**Comment by crisbeto** [comment]
We talked about how to handle this at our last team meeting. We decided against slicing the array before passing it to the constructor, because it would mean that the majority of users would incur the performance cost without much benefit. Instead, we decided to make the array immutable in development mode so that you'd get an error if you tried to modify it incorrectly. See https://github.com/angular/angular/pull/37645.
